README.md
subnettools
===========

subnet file tool !
IPv4/IPv6 subnet calculator, CSV route file manipulation and modification tool

FEATURES
========
Subnettools is a sotfware intending to help network engineers manipulate large route tables and extract
information from it; subnettools has :
- native IPv6 & IPv4 support
- IPv4 & IPv6 address information (known subnet membership, decoding of embedded IPs like Teredo)
- Powerful pattern matching engine
- route CSV files, bgp route CSV files, IPAM csv files support 
- CSV filtering, comparing, conversion from cisco etc...

Subnet arithmetic
-----------------
- relation IP1 IP2    : prints a relationship between IP1 and IP2
- split S, \<l1,l2,..\> : split subnet S l1 times, the result l2 times, and so on..
- split2 S, \<m1,m2,..\>: split subnet S with mask m1, then m2, and so on...
- removesub TYPE O1 S1: remove Subnet S from Object O1; if TYPE=file O1=ile, if TYPE=subnet 01=subnet
- ipinfo IP|all|IPvX  : prints information about IP, or all known subnets (all, IPv4 or IPv6)

Route file simplification
-------------------------
- sort FILE1          : sort CSV FILE1
- sortby name file    : sort CSV file by (prefix|gw|mask), prefix is always a tie-breaker
- sortby help	      : print available sort options
- subnetagg FILE1     : sort and aggregate subnets in CSV FILE1; GW is not checked
- routeagg  FILE1     : sort and aggregate subnets in CSV FILE1; GW is checked
- routesimplify1 FILE : simplify CSV subnet file FILE; duplicate or included networks are removed
- routesimplify2 FILE : simplify CSV subnet file FILE; prints redundant routes that can be removed

Route file comparison
---------------------
- compare FILE1 FILE2 : compare FILE1 & FILE2, printing subnets in FILE1 with relation in FILE2
- missing FILE1 FILE2 : prints subnets from FILE1 that are not covered by FILE2; GW is not checked
- uniq FILE1 FILE2    : prints unique subnets from FILE1 and FILE2
- common FILE1 FILE2  : merge CSV subnet files FILE1 & FILE2; prints common routes only; GW isn't checked
- addfiles FILE1 FILE2: merge CSV subnet files FILE1 & FILE2; prints the sum of both files
- grep FILE prefix    : grep FILE for prefix/mask
- filter FILE EXPR    : grep netcsv   FILE using regexp EXPR
- filter help         : prints help about bgp filters
- bgpfilter FILE EXPR : grep bgp_file FILE using regexp EXPR
- bgpfilter help      : prints help about bgp filters

IPAM tools
----------
- ipam <IPAM> FILE1   : print FILE1 subnets with comment extracted from IPAM (legacy)
- ipamfilter <IPAM> E : filter IPAM using regexp E
- ipamprint<IPAM>     : print <IPAM>; use option -ipamea to select Extended Attributes
- getea <IPAM> FILE   : print FILE with Extended Attributes retrieved from IPAM

ipam structure must be described in st.conf
ipam_prefix_field    : IPAM CSV header field describing the prefix
ipam_mask            : IPAM CSV header field describing the mask
ipam_comment1        : IPAM CSV header field describing comment
ipam_comment2        : IPAM CSV header field describing comment (optional)
ipam_delim           : IPAM CSV delimitor (like ',' or ';')
ipam_ea              : IPAM Extended Attributes to collect (prefix, mask and comment
by default)

if your IPAM may contain delim chars inside comments, usually comments are inside
quotes or double quotes (" or ')
you can describe that with:
ipam_comment_delim   : IPAM CSV comment delimitor
ipam_comment_delim_escape : IPAM CSV comment delimitor escape

Miscellaneous route file tools
------------------------------
- print FILE1         : just read & print FILE1; use a -fmt FMT to print CSV fields you want
- bgpprint FILE1      : just read & print BGP FILE1; use a -fmt FMT to print CSV fields you want
- sum IPv4FILE        : get total number of hosts included in the list of subnets
- sum IPv6FILE        : get total number of /64 subnets included

BGP route file tools
--------------------
- bgpcmp file1 file2  : show what changed in BGP file file1 & file2
- bgpsortby name file : sort BGP file by (prefix|gw|MED|LOCALPREF), prefix is always a tie-breaker
- bgpsortby help	    : print available sort options

IP route to CSV converters
--------------------------
- convert PARSER FILE1: convert FILE1 to csv using parser PARSER
- convert help        : use 'subnet_tool convert help' for available parsers

subnettools route FILE format is a CSV where each line represent a route ; a route is
-* a subnet
-* a subnet mask
-* a gateway
-* a device
-* a comment 
-* any number of optional Extended Attributes (EA)
- delimitors, name of the fields describing (prefix, mask, gw, dev, comment) are fully configurable
- subnettools output format is configurable ; you can configure a FMT (%I %m %D %G %C)
- subnettools has a default config file (st.conf)
- subnettools has a debug system (subnettools -D help), mostly helped me to create it
- subnettools has a small regression testing suite

LIMITS
=======
- file size : more than enough (aggretated a 8millions line CSV in 2-3minutes)
- line size : 1024 (subnettools truncates too long lines)
- comment size : 256 bytes (not really, but st_printf limits string to 256 chars)
- name of CSV fields : 32 bytes (OK unless you use stupids name to describe a prefix, mask, comment... french longest word is 26 bytes anyway, anticonstitutionnellement)
- max number of delimiters : 30 (why would you want more?)
- device name : 32 bytes (should be enough, but could be increased, TenGigabitethernet2/2/0.4121  is 29 bytes) 

it should work OK today for most functions

Input CSV format
================
When working on CSV route files, please note the following
- Input routes files SHOULD have a CSV header describing its structure (prefix, mask, GW, comment, etc...)
- Input routes files without a CSV header are assumed to be : prefix;mask;GW;comment or prefix;mask;comment
- default routes CSV header is "prefix;mask;device;GW;comment"
- CSV header can be changed by using the configuration file (subnettools confdesc for more info)
- IPAM CSV header MUST be described in the configuration file
- BGP routes files MUST have a CSV header

- some commands can take <stdin> as input : sort, sortby, bgpsortby, print, ipam, filter, bgpfilter

OUTPUT FMT
==========
The output of subnettools command working on route files can be dynamically modified by supplying an output FMT
Some theory first, next some examples :)
- %I  : the prefix
- %N  : the network address of the prefix
- %B  : the 'broadcast' address of the prefix (last subnet IP, since IPv6 has no broadcast)
- %G  : the gateway (next-hop)
- %m  : the subnet mask (prefix length notation)
- %M  : for IPv4, the mask in Dotted Decimal Notation; for IPv6 I DO REFUSE and print prefix length
- %D  : the device
- %C  : the comment
- %On : (where n is a digit) : Extensible Attribute 'n'
- %O#c: prints all Extensible Attributes, separated by char 'c'
- %U  : upper subnet of %I (lower subnet of 10.1.3.0/24 is 10.1.2.0/24)
- %L  : lower subnet 
- %P  : prefix/mask
%I, %N, %B and %G MAY be followed by a IPv6 address compression level (0, 1, 2, 3)

- compression level 0 : no compression at all
- compression level 1 : remove leading 0
- compression level 2 : remove consecutives 16bits blocks of zero
- compression level 3 : level 2 + print IPv4 Mapped & IPv4 compatible address in mixed IPv4 / IPv6 format
- compression level 4 : bitmask printing

The character % MAY be followed by a field width (see printf man pages); this can help to align the results vertically, but please note in case width is smaller than ouptut string, it WILL NOT be truncated

EXAMPLE 1 (use fmt to convert to Cisco static routes):
------------------------------------------------------
	[etienne@ARODEF subnet_tools]$ ./subnet-tools -fmt "ip route %I %M %G name %C" print regtest/route_aggipv4
	ip route prefix mask GW name comment
	ip route 10.1.0.0 255.255.255.0 192.168.1.1 name test1
	ip route 10.1.1.0 255.255.255.0 192.168.1.1 name test1
	ip route 10.1.2.0 255.255.255.0 192.168.1.1 name test1
	ip route 10.1.3.0 255.255.255.0 192.168.1.2 name test1
	ip route 10.1.4.0 255.255.255.0 192.168.1.2 name test1
	ip route 10.1.5.0 255.255.255.0 192.168.1.2 name test1
	ip route 10.1.6.0 255.255.255.0 192.168.1.2 name test1
	ip route 10.1.7.0 255.255.255.0 192.168.1.2 name test1


EXAMPLE 2 (using field width to align colons):
----------------------------------------------

	etienne@debian:~/st$ ./subnet-tools -fmt "%-16I;%-3m;%-10D;%-32G" route regtest/route_aggipv6-2 
	2000:1::        ;32 ;eth0/1    ;fe80::254                      
	2001:db4::      ;30 ;eth0/1    ;fe80::251                      
	2001:db8::      ;32 ;eth0/2    ;fe80::251                      
	2001:db9::      ;32 ;eth0/1    ;fe80::251                       
	2001:dba::      ;31 ;eth0/1    ;fe80::251                  

EXAMPLE3 (converting subnets to address ranges)
-----------------------------------------------
[etienne@ARODEF subnet_tools]$ ./subnet-tools -fmt "%13I/%2m [%N-%B]" sort regtest/aggipv4 
   
     1.1.1.0/23  [1.1.0.0-1.1.1.255]
     10.0.0.0/24 [10.0.0.0-10.0.0.255]
     10.0.1.0/24 [10.0.1.0-10.0.1.255]
     10.0.2.0/23 [10.0.2.0-10.0.3.255]
     10.0.4.0/22 [10.0.4.0-10.0.7.255]

EXAMPLE 4 (playing with IPv6 address compression)
-------------------------------------------------

	etienne@debian:~/st$ ./subnet-tools echo "%I3" 2001::1
 	2001::1
	etienne@debian:~/st$ ./subnet-tools echo "%I1" 2001::1 
	2001:0:0:0:0:0:0:1
	etienne@debian:~/st$ ./subnet-tools echo "%I0" 2001::1 
	2001:0000:0000:0000:0000:0000:0000:0001
	etienne@debian:~/st$ ./subnet-tools echo "%I2" ::ffff:10.1.1.1
	::ffff:a01:101
	etienne@debian:~/st$ ./subnet-tools echo "%I3" ::ffff:10.1.1.1 
	::ffff:10.1.1.1

IP information
==============
subnettools ipinfo 'subnet' can give the following informations about 'subnet'
- network address, last subnet address
- known prefix membership (IPv6 global addresses, RFC 1918 private subnet, multicast, etc...)
Decoding of following addresses :
- Teredo
- IPv6 multicast (including embedded RP)
- 6to4
- ISATAP link-local address
- IPv4 glop
- rfc 6052
- EUI-64 Interface ID

Example :
---------

	[etienne@ARODEF subnet_tools]$ ./subnet-tools ipinfo 2001:0000:4136:e378:8000:63bf:3fff:fdd2 
	IP version : 6
	Network Address : 2001:0:4136:e378:8000:63bf:3fff:fdd2/128
	Address   Range : 2001:0:4136:e378:8000:63bf:3fff:fdd2 - 2001:0:4136:e378:8000:63bf:3fff:fdd2
	Previous subnet : 2001:0:4136:e378:8000:63bf:3fff:fdd1/128
	Next     subnet : 2001:0:4136:e378:8000:63bf:3fff:fdd3/128
	IPv6 rfc4380 Teredo
	Teredo server : 65.54.227.120
	Client IP     : 192.0.2.45
	UDP port      : 40000

	[etienne@ARODEF subnet_tools]$ ./subnet-tools ipinfo FF72:340:2001:DB8:BEEF:FEED::32
	IP version : 6
	Network Address : ff72:340:2001:db8:beef:feed:0:32/128

CHANGELOG
v1.6 (unreleased)
=================
-New features
-- supports IPAM comments containing the delim char (1.1.1.1,24,"comment,is,good")


v1.5 (2018 refresh)
===================

-New features
-- Case insensitive pattern matching, use '%' special char for filter :
[etienne@ARODEF subnet_tools]$ ./subnet-tools -ea EA-Site ipamfilter regtest/ipam-test 'EA-Site%.*sain.*'
prefix;EA-Site;
10.128.0.0/16;Saint Denis;
10.128.1.0/24;Saint Denis;
-- scanf, fscanf can use user-defined format taken from '-fmt' options
-- ./subnet-tools -fmt "%O0;%O1\n" scanf "1.1.1.1        y" "%I *%c"
1.1.1.1;y
-- new command 'limits' to print subnettools known limits


- Internal changes
-- optimized read_csv when delims is only ONE (common case)
-- write a new file opening/reading framework (st_getline_truncate) to read lines from text file and truncate the line to a maximum allowed length
-- rewrite of string2addr family functions, major speedups
-- rewrite major parts of st_scanf pattern matching engine
-- st_scanf mutliplier handling (+, *, ?) was trying to be generic, dividing it into 3 separates cases helps
readability a lot, speeds up and simplify code
-- st_scanf error handling was simplified and unified

v1.4
====
-New features
-- bitmask printing (with compression 4, ie use '%I4' in fmt)
-- Extensible Attribute attached to routes, can be also filtered

- Miscellaneous new commands
-- 'compare'  now returns all relations (INCLUDED, INCLUDES, EQUALS)
-- subnetcmp before after : prints all changes that occurs between 'before' and 'after'; you should use typically after you change your routing

- new IPAM commands and IPAM infrastructure
-- ipamprint  (Example : ./subnettools -ea EA-Site,EA-Vlan,comment ipamprint ipam.csv )
-- ipamfilter (Example : ./subnettools -ea EA-Site,EA-Vlan,comment ipamfilter ipam.csv 'EA-Vlan<2000&EA-Site=Acheres')
-- getea (Example : ./subnettools -EA-Site,comment getea ipam.csv toto)

- Internal changes
-- st_malloc infrastructure to track memory leaks
-- struct route a no more comment field, just Extensible Attributes
-- many modification to CSV helpers, to load files with varying number of Extensible Attributes
-- checked by checkpatch.pl from Linux

v1.3
====
-New features
-- filter & bgpfilter commands (filter on GW, prefix, AS_PATH, etc... in any logical combination)
-- add 'uniq' commands, prints routes from FILE1 & FILE2 that have no relation
-- remove diff command, that never worked properly
-- bgpprint according to a format!

-Internal changes
-- bug fixes (didnt properly scan 'sh ip bgp' , because some subnets are ... classfull)
-- rewritten string2mask low level function
-- improved error reporting (with -V, -VV), debug message levels are more consistent

v1.2
====
-New features
--ecmp options to print all routes in case the source files has multiples NH to same destination
--many fixes in converters
--sortby, bgpsorty commands to sort BGP routes, subnet routes
--revamped help (in categories)
--can compile on freebsd now

v1.1
====
New features :
- converters now supports '-rt' option to set comment to route type (ospf, bgp, static ....)
- converter for show route on ASA
-BGP tools :
-- converter to load a BGP table
-- bgpcmp command to show differences between a table, after some UPDATES

v1.0
====
- New features
-- network split
-Internal changes
-- lot of cleanups, scanf work
-- all converters converted to st_scanf

v0.7
====
-Internal changes
-- st_scanf pattern matching engine
-- few visible changes for the end user, except a new 'scanf' command
-- rewriting and hardening of low -level IPv4/IPv6 parsing functions

v0.6
====
- New features
--ipinfo :
subnettools ipinfo 'subnet' can give the following informations about 'subnet'
- network address, last subnet address
- known prefix membership (IPv6 global addresses, RFC 1918 private subnet, multicast, etc...)
Decoding of following addresses :
- Teredo
- IPv6 multicast (including embedded RP)
- 6to4
- ISATAP link-local address
- IPv4 glop
- rfc 6052
- EUI-64 Interface ID

v0.4
====
- Internal changes
-- dynamic output fmt


ipinfo.c
/*
 * IP address information
 *
 * Copyright (C) 2014 Etienne Basset <etienne POINT basset AT ensta POINT org>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License
 * as published by the Free Software Foundation.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "iptools.h"
#include "debug.h"
#include "st_printf.h"
#include "ipinfo.h"

const struct subnet class_a = { .ip_ver = IPV4_A, .ip = 0, .mask = 1};
const struct subnet class_b = { .ip_ver = IPV4_A, .ip = (2 << 30), .mask = 2};
const struct subnet class_c = { .ip_ver = IPV4_A, .ip = (6 << 29), .mask = 3};
const struct subnet class_d = { .ip_ver = IPV4_A, .ip = (14 << 28), .mask = 4};
const struct subnet class_e = { .ip_ver = IPV4_A, .ip = (15 << 28), .mask = 4};

#define S_IPV4_CONST(DIGIT1, DIGIT2, __MASK)  { .ip_ver = IPV4_A, .ip = (DIGIT1 << 24) + (DIGIT2 << 16), .mask = __MASK }
const struct subnet ipv4_broadcast	= { .ip_ver = IPV4_A, .ip = -1, .mask = 32};
const struct subnet ipv4_default	= S_IPV4_CONST(0,  0,   0);
const struct subnet ipv4_unspecified	= S_IPV4_CONST(0,  0,   32);
const struct subnet ipv4_rfc1918_1	= S_IPV4_CONST(10,  0,   8);
const struct subnet ipv4_rfc1918_2	= S_IPV4_CONST(172, 16,  12);
const struct subnet ipv4_rfc1918_3	= S_IPV4_CONST(192, 168, 16);
const struct subnet ipv4_loopback	= S_IPV4_CONST(127, 0,   8);
const struct subnet ipv4_rfc3927_ll	= S_IPV4_CONST(169, 254, 16);
const struct subnet ipv4_rfc6598_sasr	= S_IPV4_CONST(100, 64,  10);
const struct subnet ipv4_bench		= S_IPV4_CONST(198, 18,  15);
#define S_IPV4_CONST3(DIGIT1, DIGIT2, DIGIT3,__MASK)  { .ip_ver = 4, .ip = (DIGIT1 << 24) + (DIGIT2 << 16) + (DIGIT3 << 8), .mask = __MASK }
const struct subnet ipv4_test1		= S_IPV4_CONST3(192, 0, 2, 24);
const struct subnet ipv4_test2		= S_IPV4_CONST3(198, 51, 100, 24);
const struct subnet ipv4_test3		= S_IPV4_CONST3(203, 0, 113, 24);

const struct subnet ipv4_mcast_ll	= S_IPV4_CONST(224,0, 8);
const struct subnet ipv4_mcast_ssm	= S_IPV4_CONST(232,0, 8);
const struct subnet ipv4_mcast_glop	= S_IPV4_CONST(233,0, 8);
const struct subnet ipv4_mcast_site	= S_IPV4_CONST(239,0, 8);

#define S_IPV6_CONST(DIGIT1, DIGIT2, __MASK)  { .ip_ver = 6, block(.ip6, 0) = DIGIT1, block(.ip6, 1) = DIGIT2, .mask = __MASK }
const struct subnet ipv6_default	= S_IPV6_CONST(0x0000, 0, 0);
const struct subnet ipv6_unspecified	= S_IPV6_CONST(0x0000, 0, 128);
const struct subnet ipv6_global		= S_IPV6_CONST(0x2000, 0, 3);
const struct subnet ipv6_ula		= S_IPV6_CONST(0xfc00, 0, 7);
const struct subnet ipv6_sitelocal	= S_IPV6_CONST(0xfec0, 0, 10);
const struct subnet ipv6_linklocal	= S_IPV6_CONST(0xfe80, 0, 10);
const struct subnet ipv6_multicast	= S_IPV6_CONST(0xff00, 0, 8);
const struct subnet ipv6_6to4		= S_IPV6_CONST(0x2002, 0, 16);
const struct subnet ipv6_rfc4380_teredo = S_IPV6_CONST(0x2001, 0, 32);
const struct subnet ipv6_rfc3849_doc	= S_IPV6_CONST(0x2001, 0x0DB8, 32);
const struct subnet ipv6_rfc6052_pat 	= S_IPV6_CONST(0x0064, 0xff9b, 96);
const struct subnet ipv6_isatap_priv_ll	= {.ip_ver = 6, block(.ip6, 0) = 0xFE80, block(.ip6, 5) = 0x5EFE, .mask = 96};
const struct subnet ipv6_isatap_pub_ll	= {.ip_ver = 6, block(.ip6, 0) = 0xFE80, block(.ip6, 4) = 0x0200, block(.ip6, 5) = 0x5EFE, .mask = 96};
const struct subnet ipv6_mapped_ipv4	= {.ip_ver = 6, block(.ip6, 5) = 0xFFFF, .mask = 96}; /* ::FFFF:/96 */
const struct subnet ipv6_mcast_sn	= {.ip_ver = 6, block(.ip6, 0) = 0xFF02, block(.ip6, 5) = 0x1, block(.ip6, 6) = 0xFF00, .mask = 104};
const struct subnet ipv6_compat_ipv4	= {.ip_ver = 6, .mask = 96}; /* ::/96 */
const struct subnet ipv6_loopback 	= {.ip_ver = 6, block(.ip6, 7) = 1, .mask = 128}; /* ::1/128 */

static void decode_6to4(FILE *out, const struct subnet *s) {
	fprintf(out, "6to4 IPv4 destination address : %d.%d.%d.%d\n", block(s->ip6, 1) >> 8, block(s->ip6, 1) & 0xFF,
			block(s->ip6, 2) >> 8, block(s->ip6, 2) & 0xFF);
}

static void decode_teredo(FILE *out, const struct subnet *s) {
	fprintf(out, "Teredo server : %d.%d.%d.%d\n", block(s->ip6, 2) >> 8, block(s->ip6, 2) & 0xFF,
			block(s->ip6, 3) >> 8, block(s->ip6, 3) & 0xFF);
	fprintf(out, "Client IP     : %d.%d.%d.%d\n", (block(s->ip6, 6) >> 8) ^ 0xFF , (block(s->ip6, 6) & 0xFF) ^ 0xFF,
			(block(s->ip6, 7) >> 8) ^ 0xFF, (block(s->ip6, 7) & 0xFF) ^ 0xFF);
	fprintf(out, "UDP port      : %d\n", block(s->ip6, 5) ^ 0xFFFF);
}

static void decode_rfc6052(FILE *out, const struct subnet *s) {
	fprintf(out, "IPv4-Embedded IPv6 address : 64:ff9b::%d.%d.%d.%d\n", block(s->ip6, 6) >> 8, block(s->ip6, 6) & 0xFF,
			block(s->ip6, 7) >> 8, block(s->ip6, 7) & 0xFF);
}

static void decode_isatap_ll(FILE *out, const struct subnet *s) {
	fprintf(out, "ISATAP IPv4 destination address : %d.%d.%d.%d\n", block(s->ip6, 6) >> 8, block(s->ip6, 6) & 0xFF,
			block(s->ip6, 7) >> 8, block(s->ip6, 7) & 0xFF);
}

static void decode_ipv4_multicast(FILE *out, const struct subnet *s) {
	int i, res;
	const struct known_subnet_desc *k = ipv4_mcast_known_subnets;
	int found_mask, found_i;

	found_mask = -1;
	for (i = 0; ;i++) {
		if (k[i].s == NULL)
			break;
		res = subnet_compare(s, k[i].s);
		st_debug(ADDRCOMP, 5, "%P against %P\n", *s, *k[i].s);
		if (res == EQUALS || (res == INCLUDED && k[i].always_print != -1)) {
			if ((int)k[i].s->mask >= found_mask) {
				found_mask = k[i].s->mask;
				found_i = i;
			}
		}
	}
	if (found_mask >= 0) {
		fprintf(out, "%s\n", k[found_i].desc);
		if (k[found_i].decode_more)
			k[found_i].decode_more(out, s);
	}
}

static void decode_ipv6_embedded_rp(FILE *out, const struct subnet *s) {
	struct subnet rp;
	int rp_id;
	int Plen;
	u32 group_id;
	int i, j;

	memset(&rp, 0, sizeof(rp));
	rp.ip_ver = IPV6_A;
	rp.mask   = 128;
	rp_id     = (block(s->ip6, 1) >> 8) & 0xF;
	Plen      = block(s->ip6, 1) & 0xFF;
	if (Plen > 64) {
		fprintf(out, "Invalid Plen %x, MUST not be greater than 64\n", Plen);
		return;
	}
	group_id = block(s->ip6, 6) * (1 << 16) + block(s->ip6, 7);
	/* copying Plen bits of s into rp */
	for (i = 0; i < Plen / 16; i++)
		set_block(rp.ip6, i, block(s->ip6, i + 2));
	for (j = 0 ; j < Plen % 16; j++)
		block_OR(rp.ip6, i, block(s->ip6, i + 2) & (1 << (15 - j)));
	set_block(rp.ip6, 7, rp_id);
	st_fprintf(out, "Embedded RP Address : %I\n", rp);
	fprintf(out, "32-bit group id 0x%x [%d]\n", group_id, group_id);
}

/* solicitted node address */
static void decode_ipv6_multicast_sn(FILE *out, const struct subnet *s) {
	fprintf(out, "Sollicited Address for any address like : XX:XX:XX:XX:XX:XX:X%02x:%x\n", 
			block(s->ip6, 6) & 0xFF, block(s->ip6, 7));
}

/* generic IPv6 multicast decoding */
static void decode_ipv6_multicast(FILE *out, const struct subnet *s) {
	int scope, flag;
	scope = block(s->ip6, 0) & 0xf;
	flag  = (block(s->ip6, 0) >> 4) & 0xf;

	fprintf(out, "Scope : ");
	switch (scope) {
		case 1:
			fprintf(out, "Host Local\n");
			break;
		case 2:
			fprintf(out, "Link Local\n");
			break;
		case 5:
			fprintf(out, "Site Local\n");
			break;
		case 8:
			fprintf(out, "Organisation Local\n");
			break;
		case 14:
			fprintf(out, "Global\n");
			break;
		default:
			fprintf(out, "Invalid : %d\n", scope);
			break;
	}
	fprintf(out, "Flags : %d\n", flag);
	if (flag & 4)
		fprintf(out, "R=1, Embedded RP\n");
	else
		fprintf(out, "R=0, No Embedded RP\n");
	if (flag & 2)
		fprintf(out, "P=1, based on network prefix\n");
	else
		fprintf(out, "P=0, not based on network prefix\n");
	if (flag & 1)
		fprintf(out, "T=1, dynamically assigned prefix\n");
	else
		fprintf(out, "T=0, well-known multicast address\n");
	if (flag == 7)
		decode_ipv6_embedded_rp(out, s);
}

const struct known_subnet_desc ipv4_known_subnets[] = {
	{&ipv4_default,		"IPv4 default address", -1},
	{&ipv4_broadcast,	"IPv4 broadcast address", -1},
	{&class_d,		"IPv4 multicast address", 1, &decode_ipv4_multicast},
	{&ipv4_unspecified,	"IPv4 unspecified address"},
	{&ipv4_rfc1918_1,	"IPv4 Private address (rfc1918)"},
	{&ipv4_rfc1918_2,	"IPv4 Private address (rfc1918)"},
	{&ipv4_rfc1918_3,	"IPv4 Private address (rfc1918)"},
	{&ipv4_loopback,	"IPv4 loopback address"},
	{&ipv4_rfc3927_ll,	"IPv4 link-local address (rfc6890)"},
	{&ipv4_rfc6598_sasr,	"IPv4 private Shared Address Space (rfc6598)"},
	{&ipv4_test1,		"IPv4 Test & doc NET (rfc5737)"},
	{&ipv4_test2,		"IPv4 Test & doc NET (rfc5737)"},
	{&ipv4_test3,		"IPv4 Test & doc NET (rfc5737)"},
	{&ipv4_bench,		"IPv4 benchmark test (rfc2544)"},
	{NULL, NULL}
};

const struct known_subnet_desc ipv6_known_subnets[] = {
	{&ipv6_default,		"IPv6 default address", -1},
	{&ipv6_unspecified,	"IPv6 unspecified address"},
	{&ipv6_global,		"IPv6 global address"},
	{&ipv6_ula,		"IPv6 Unique Local Address"},
	{&ipv6_sitelocal,	"IPv6 Site Local address (deprecated)"},
	{&ipv6_linklocal,	"IPv6 link-local address"},
	{&ipv6_multicast,	"IPv6 multicast address", 1, &decode_ipv6_multicast},
	{&ipv6_mcast_sn,	"IPv6 multicast Solicited Node Address", 1, &decode_ipv6_multicast_sn},
	{&ipv6_6to4,		"IPv6 6to4", 1, &decode_6to4},
	{&ipv6_rfc4380_teredo,	"IPv6 rfc4380 Teredo", 1, &decode_teredo},
	{&ipv6_rfc3849_doc,	"IPv6 rfc3849 Documentation-reserved addresses"},
	{&ipv6_rfc6052_pat,	"IPv6 rfc6052 protocol address translation", 1, &decode_rfc6052},
	{&ipv6_isatap_priv_ll,	"IPv6 ISATAP link-local address", 1, &decode_isatap_ll},
	{&ipv6_isatap_pub_ll,	"IPv6 ISATAP link-local address", 1, &decode_isatap_ll},
	{&ipv6_mapped_ipv4,	"IPv6 Mapped-IPv4 address"},
	{&ipv6_compat_ipv4,	"IPv6 Compat-IPv4 address (Deprecated)"},
	{&ipv6_loopback,	"IPv6 Loopback Address"},
	{NULL, NULL}
};

static void decode_mcast_glop(FILE *out, const struct subnet *s) {
	fprintf(out, "Glop AS : %d\n", (s->ip >> 8) & 0xFFFF);
}

const struct known_subnet_desc ipv4_mcast_known_subnets[] = {
	{&ipv4_mcast_ll,	"IPv4 Link-Local Multicast Addresses"},
	{&ipv4_mcast_ssm,	"IPv4 Source Specific Multicast Addresses"},
	{&ipv4_mcast_glop,	"IPv4 Glop Multicast Addresses (rfc2770)", 1, &decode_mcast_glop},
	{&ipv4_mcast_site,	"IPv4 Site Local (private) Multicast Addresses"},
	{NULL, NULL}
};

/*
 * try to find a known subnet where s is included
 */
static void fprint_ip_membership(FILE *out, const struct subnet *s) {
	int i, res;
	int found_i, found_mask;
	const struct known_subnet_desc *k;

	if (s->ip_ver == IPV4_A)
		k = ipv4_known_subnets;
	else if (s->ip_ver == IPV6_A)
		k = ipv6_known_subnets;
	else
		return;
	found_mask = -1;
	for (i = 0; ;i++) {
		if (k[i].s == NULL)
			break;
		res = subnet_compare(s, k[i].s);
		st_debug(ADDRCOMP, 5, "%P against %P\n", *s, *k[i].s);
		if (res == EQUALS || (res == INCLUDED && k[i].always_print != -1)) {
			if ((int)k[i].s->mask >= found_mask) {
				found_mask = k[i].s->mask;
				found_i = i;
			}
		}
	}
	if (found_mask >= 0) {
		fprintf(out, "%s\n", k[found_i].desc);
		if (k[found_i].decode_more)
			k[found_i].decode_more(out, s);
	}
}

char ipv4_get_class(const struct subnet *s) {
	if ((s->ip >> 31) == 0)
		return 'a';
	if ((s->ip >> 30) == 2)
		return 'b';
	if ((s->ip >> 29) == 6)
		return 'c';
	if ((s->ip >> 28) == 14)
		return 'd';
	if ((s->ip >> 28) == 15)
		return 'e';
	return 0;
}

/* IPv4 specific information : classfull info */
static void fprint_ipv4_info(FILE *out, const struct subnet *subnet) {
	char c;

	c = ipv4_get_class(subnet);
	st_fprintf(out, "Mask length : %m DDN: %M\n", *subnet, *subnet);
	fprintf(out, "Classfull info : class %c\n", c);
	fprint_ip_membership(out, subnet);
}

/* IPv6 specific information EUI-64 etc */
static void fprint_ipv6_info(FILE *out, const struct subnet *s) {
	/*int ul_bit; */
	unsigned short middle;
	unsigned char mac[8];

	fprint_ip_membership(out, s);
	/* if is_global or is_link_local or ULA */
	if (ipv6_is_global(s->ip6) || ipv6_is_ula(s->ip6) || ipv6_is_link_local(s->ip6)) {
		/* ul_bit  = (s->ip6.n16[4] >> 8) & 0x02; */
		middle  = (block(s->ip6, 6) >> 8) & 0xFF;
		middle += (block(s->ip6, 5) & 0xFF) << 8;
		if (middle == 0xFFFE) {
			fprintf(out, "Probably in EUI-64 format\n");
			mac[0] = (block(s->ip6, 4) >> 8) ^ 0x02;
			mac[1] = (block(s->ip6, 4) & 0xFF);
			mac[2] = (block(s->ip6, 5) >> 8);
			mac[3] = (block(s->ip6, 6) & 0xFF);
			mac[4] = (block(s->ip6, 7) >> 8);
			mac[5] = (block(s->ip6, 7) & 0xFF);
			fprintf(out, "MAC address : %02x:%02x:%02x:%02x:%02x:%02x\n", mac[0], mac[1], mac[2],
					mac[3], mac[4], mac[5]);
		} else {
			fprintf(out, "Not in EUI-64 format\n");
		}
	}
}
/* general information about the subnet */
void fprint_ip_info(FILE *out, const struct subnet *subnet) {
	fprintf(out, "IP version : %d\n", subnet->ip_ver);
	st_fprintf(out, "Network Address : %N/%m\n", *subnet, *subnet);
	st_fprintf(out, "Address   Range : %N - %B\n", *subnet, *subnet);
	st_fprintf(out, "Previous subnet : %L/%m\n", *subnet, *subnet);
	st_fprintf(out, "Next     subnet : %U/%m\n", *subnet, *subnet);
	if (subnet->ip_ver == IPV4_A)
		fprint_ipv4_info(out, subnet);
	if (subnet->ip_ver == IPV6_A)
		fprint_ipv6_info(out, subnet);
}

static void fprint_ipv4_known_mcast_subnets(FILE *out) {
	const struct known_subnet_desc *k;
	int i;

	k = ipv4_mcast_known_subnets;
	for (i = 0; ; i++) {
		if (k[i].desc == NULL)
			break;
		st_fprintf(out, "%-18P : [%-15N - %-16B] -- %s\n", *k[i].s, *k[i].s, *k[i].s, k[i].desc);
	}
}

void fprint_ipv4_known_subnets(FILE *out) {
	const struct known_subnet_desc *k = ipv4_known_subnets;
	int i;

	for (i = 0; ; i++) {
		if (k[i].desc == NULL)
			break;
		st_fprintf(out, "%-18P : [%-15N - %-16B] -- %s\n", *k[i].s, *k[i].s, *k[i].s, k[i].desc);
	}
	fprint_ipv4_known_mcast_subnets(out);
}

void fprint_ipv6_known_subnets(FILE *out) {
	const struct known_subnet_desc *k = ipv6_known_subnets;
	int i;

	for (i = 0; ; i++) {
		if (k[i].desc == NULL)
			break;
		st_fprintf(out, "%-21P : [%-18N - %-40B] -- %s\n", *k[i].s, *k[i].s, *k[i].s, k[i].desc);
	}
}
